RFS/src/test/java/org/opensearch/migrations/bulkload/common/LuceneDocumentsReaderTest.java

```diff
@@ -277,6 +278,98 @@ protected DirectoryReader getReader() {
+        );
+        Path luceneDir = unpacker.unpack();
+
+        // Use the LuceneDocumentsReader to get the documents
+        var reader = LuceneDocumentsReader.getFactory(sourceResourceProvider).apply(luceneDir);
+
+
+        for (int i = 0; i < documentStartingIndices.size(); i++) {
+            Flux<RfsLuceneDocument> documents = reader.readDocuments(0, documentStartingIndices.get(i))
+                    .sort(Comparator.comparing(doc -> doc.id)); // Sort for consistent order given LuceneDocumentsReader may interleave
+
+            var verifier = StepVerifier.create(documents);
+            var expectedDocumentIds = documentIds.get(i);
+            expectedDocumentIds.forEach(id -> {
+                    verifier.expectNextMatches(doc -> {
+                        Assertions.assertEquals(id, doc.id);
+                        return true;
+                    });
+            });
+
+            verifier.expectComplete().verify();
+        }
+    }
+    @Test
+    public void ReadDocumentsStartingFromCheckpointForManySegments_AsExpected() throws Exception {
+        // This snapshot has three segments, each with only a single document.
+        var snapshot = TestResources.SNAPSHOT_ES_6_8;
+        var version = Version.fromString("ES 6.8");
+        List<List<String>> documentIds = List.of(
+                List.of("complexdoc", "unchangeddoc", "updateddoc"),
+                List.of("unchangeddoc", "updateddoc"),
+                List.of("unchangeddoc"));
+
+        final var repo = new FileSystemRepo(snapshot.dir);
+        var sourceResourceProvider = ClusterProviderRegistry.getSnapshotReader(version, repo);
+        DefaultSourceRepoAccessor repoAccessor = new DefaultSourceRepoAccessor(repo);
+
+        final ShardMetadata shardMetadata = sourceResourceProvider.getShardMetadata().fromRepo(snapshot.name, "test_updates_deletes", 0);
+
+        SnapshotShardUnpacker unpacker = new SnapshotShardUnpacker(
+                repoAccessor,
+                tempDirectory,
+                shardMetadata,
+                Integer.MAX_VALUE
+        );
+        Path luceneDir = unpacker.unpack();
+
+        // Use the LuceneDocumentsReader to get the documents
+        var reader = LuceneDocumentsReader.getFactory(sourceResourceProvider).apply(luceneDir);
+
+
+        for (int i = 0; i < documentIds.size(); i++) {
+            Flux<RfsLuceneDocument> documents = reader.readDocuments(i, 0)
+                    .sort(Comparator.comparing(doc -> doc.id)); // Sort for consistent order given LuceneDocumentsReader may interleave
+
+            var verifier = StepVerifier.create(documents);
+            var expectedDocumentIds = documentIds.get(i);
+            expectedDocumentIds.forEach(id -> {
+                verifier.expectNextMatches(doc -> {
+                    Assertions.assertEquals(id, doc.id);
+                    return true;
+                });
+            });
+
+            verifier.expectComplete().verify();
```

Comment:
This is really nice code - but you won't get full context back when it fails.  Something to consider would be to just concatenate the expected data into a list or json & then do the same for the test data.  JUnit runners are pretty good about showing diffs